keps/sig-api-machinery/2309-idl/syntax.md

```diff
@@ -0,0 +1,635 @@
+            containers: create-only list-map(value: Container),
+            struct Container {
+                name: string,
+                // ...
+            }
+        }
+
+        status: Status,
+
+        struct Status {
+
+        }
+    }
+
+    wrapper DNSLabel: string validates(pattern: "[a-z0-9]([-a-z0-9]*[a-z0-9])?", max-length: 64);
+
+    struct Volume {
+        name: DNSLabel,
+        source: Source,
+
+        // unions contain several variations, only one of which
+        // may be set at any given time.  By default, they have
+        // a tag of `type`.
+        // If a variant has no body, simply don't specify a tag.
+        union Source {
+            hostPath: HostPath,
+            struct HostPath {
+                path: string,
+            }
+            emptyDir: EmptyDir,
+            struct EmptyDir {}
+            // ...
+        }
+    }
+
+    // enums represent one choice of a series of constant string values
+    enum DNSPolicy {
+        /// resolve cluster first, without host networking
+        ClusterFirstWithHostNet,
+        /// resolve cluster first
+        ClusterFirst,
+        /// resolve in the default order
+        Default,
+        /// don't resolve DNS
+        None,
+    }
+
+    // markers are extensions that annotate fields or types with
+    // additional data that can be picked up by tooling.
+    // several markers are built-in (e.g. `deprecated`)
+    //
+    // The examples below imagine a marker that denotes that a given
+    // field actually contains a reference to another object.
+    // Other markers might include overrides for legacy Go type names,
+    // marking fields as being behind feature gates for automated linting
+    // and better documentation, etc.
+
+    struct SecretProjection {
+        // inline fields have their fields embedded in their parent object
+        // in systems where this is supported (e.g. JSON, Go).  In systems
+        // where not supported, generally just use the type name as the field
+        // name.  Only structs and unions may be inlined.
+
+        // "raw identifiers" are surrounded with backticks, and take two purposes
+        // the first is to allow fields and keys that have the same name as keywords
+        // (like below).
+        //
+        // The second is as an emergency "break glass" to violate the "field name must
+        // have an lowercase first letter" constraint.
+        @reference(group: "v1", version: "core", `kind`: "Secret")
+        _inline: LocalObjectReference,
+    }
+
+    @reference(same-namespace: true)
+    struct LocalObjectReference {
+        // fields are referenced using `.name`
+        // (this avoids parsing ambiguities with primitive type names)
+
+        @reference-part(part: .name)
+        name: string,
+    }
+
+}
+```
+
+```
+import (
+    types (
+       // group-versions are imported explicitly from KDL or CKDL files
+       {core/v1} from "k8s.io/api/core/v1";
```

Comment:
With the current psuedo-IDL language, we are able to make sure that the versions of these KDL or CKDL files are controlled via Go Modules. Will there be a way of stating a version to target for these imported APIs?

Reply:
Whoops, good catch, I think somehow this ended up with an older version of the imports path in the example, gotta fix that up.

The actual plan is

```
{core/v1} from "core/v1/types.kdl"
```

I believe it's possible to distribute non-Go files with go packaging, although I'd have to double check the exact details (it's been a little while since I looked).

That being said, it's kind-of a goal to allow decoupling this from Go's version management, if folks wish.  I'd imaging we'd also end up publishing an artifact like `kubernetes-1.Y.Z.ckdl` for each release that you could just import from as well.

Under the hood, this ends up looking a lot like the proto commands, so on the command line, you'd either do: 

```
# from a whole bunch of files on disk, imagine:
$ ls path/to/kubernetes/1.19
core/v1/types.kdl
batch/v1/types.kdl
...
$ kdlc -I path/to/kubernetes/1.19/ myfile.kdl
# from a bundle with virtual paths
$ kdlc -b path/to/kubernetes-1.19.0.ckdl myfile.kdl
```